### PR TITLE
Improve type safety for remote memory reads

### PR DESCRIPTION
**Pull request by godlygeek**
Associate a type with every field of each structure that we are able to read from the remote process, and ensure that any reads of that field return the value as its associated type.

### CODE REVIEW

**Comment by godlygeek** [comment]
I think this is a pretty big improvement. It conflicts with #108 so we'll need to rebase one after landing the other.

**Comment by pablogsal** [comment]
Let’s land this first

File: src/pystack/_pystack/process.h [hunk 1]
```diff
@@ -68,59 +68,19 @@ class AbstractProcessManager : public std::enable_shared_from_this<AbstractProce
...
-        return versionedField<T, PyRuntimeState, &python_v::py_runtime, PMD1>(py_runtime);
-    }
-
-    template<typename T, auto PMD1>
-    inline auto& versionedTypeField(const PyTypeObject& py_type) const
-    {
-        return versionedField<T, PyTypeObject, &python_v::py_type, PMD1>(py_type);
+        offset_t offset = getFieldOffset(field);
+        auto address = reinterpret_cast<const char*>(&obj) + offset;
+        return *(typename FieldPointer::Type*)(address);
```
Comment (pablogsal) [inline src/pystack/_pystack/process.h:83]: Nit: 

```suggestion
        return *reinterpret_cast<typename FieldPointer::Type*>(address);
```

File: src/pystack/_pystack/process.h [hunk 2]
```diff
@@ -68,59 +68,19 @@ class AbstractProcessManager : public std::enable_shared_from_this<AbstractProce
...
-        offset_t offset = d_py_v->*PMD0.*PMD1;
-        return (*((T*)(((char*)&py_obj) + offset)));
+        return (d_py_v->get<OffsetsStruct>().*field).offset;
     }
 
-    template<typename T, auto PMD1>
-    inline auto& versionedThreadField(const PyThreadState& py_thread) const
+    template<typename OffsetsStruct, typename FieldPointer>
+    inline typename FieldPointer::Type&
+    getField(const typename OffsetsStruct::Structure& obj, FieldPointer OffsetsStruct::*field) const
```
Comment (pablogsal) [inline src/pystack/_pystack/process.h:79]: Oh, very clever!

**Review by pablogsal — APPROVED** [review]
LGTM

I left one nit that I am correcting myself

**Comment by pablogsal** [comment]
Thanks a lo @godlygeek for working on this. This is a fantastic improvement and will make it much easier to spot mistakes and work with new versions in the future 🖤

**Pull request merged**